# Exploratory Data Analysis & Data Cleaning

In [1]:
# import requisite libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# allow inline plotting
%matplotlib inline

In [2]:
# checking all the available datasets provided in the project folder
import os
file_list = os.listdir(r"C:/Users/user/Documents/Flatiron/dsc-movie-studio-project/data")
for file_name in file_list:
    print(file_name)

bom.movie_gross.csv
imdb.name.basics.csv
imdb.title.akas.csv
imdb.title.basics.csv
imdb.title.crew.csv
imdb.title.principals.csv
imdb.title.ratings.csv
rt.movie_info.tsv
rt.reviews.tsv
tmdb.movies.csv
tn.movie_budgets.csv


In [3]:
!ls

data
EDA.ipynb
img
LICENSE
README.md
student.ipynb


### Dataset Inspection
In order to select relevant datasets and ultimately relevant features, we will inspect all the availble datasets. Specifically, we will look at the features contained in each dataset to determine whether they will be relevant to the project.

In [4]:
tn_movie_budget_df = pd.read_csv("./data/tn.movie_budgets.csv", index_col="id", header=0)
tmdb_movies_df = pd.read_csv("./data/tmdb.movies.csv", header=0)
reviews_df = pd.read_csv("./data/rt.reviews.tsv", index_col="id", header=0, delimiter="\t", encoding="latin1")
movie_info_df = pd.read_csv("./data/rt.movie_info.tsv", header=0, index_col="id", delimiter="\t")
titles_ratings_df = pd.read_csv("./data/imdb.title.ratings.csv", header=0)
titles_principals_df = pd.read_csv("./data/imdb.title.principals.csv", header=0)
titles_crew_df = pd.read_csv("./data/imdb.title.crew.csv", header=0)
bom_df = pd.read_csv("./data/bom.movie_gross.csv", header=0)
titles_aka_df = pd.read_csv("./data/imdb.title.akas.csv", header=0)
name_basics_df = pd.read_csv("./data/imdb.name.basics.csv", header=0)
titles_basics_df = pd.read_csv("./data/imdb.title.basics.csv", header=0)

In [5]:
# inspecting the bom.movie_gross dataset
bom_df.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [6]:
# inspecting the name basics dataset
name_basics_df.head()

nconst       primary_name  birth_year  death_year  \
0  nm0061671  Mary Ellen Bauder         NaN         NaN   
1  nm0061865       Joseph Bauer         NaN         NaN   
2  nm0062070         Bruce Baum         NaN         NaN   
3  nm0062195       Axel Baumann         NaN         NaN   
4  nm0062798        Pete Baxter         NaN         NaN   

                                 primary_profession  \
0         miscellaneous,production_manager,producer   
1        composer,music_department,sound_department   
2                        miscellaneous,actor,writer   
3  camera_department,cinematographer,art_department   
4  production_designer,art_department,set_decorator   

                          known_for_titles  
0  tt0837562,tt2398241,tt0844471,tt0118553  
1  tt0896534,tt6791238,tt0287072,tt1682940  
2  tt1470654,tt0363631,tt0104030,tt0102898  
3  tt0114371,tt2004304,tt1618448,tt1224387  
4  tt0452644,tt0452692,tt3458030,tt2178256

In [7]:
# inspecting the title akas dataset
titles_aka_df.head()

title_id  ordering                                    title region  \
0  tt0369610        10                            Джурасик свят     BG   
1  tt0369610        11                        Jurashikku warudo     JP   
2  tt0369610        12  Jurassic World: O Mundo dos Dinossauros     BR   
3  tt0369610        13                  O Mundo dos Dinossauros     BR   
4  tt0369610        14                           Jurassic World     FR   

  language        types   attributes  is_original_title  
0       bg          NaN          NaN                0.0  
1      NaN  imdbDisplay          NaN                0.0  
2      NaN  imdbDisplay          NaN                0.0  
3      NaN          NaN  short title                0.0  
4      NaN  imdbDisplay          NaN                0.0

In [8]:
# inspecting the title basics dataset
titles_basics_df.head()

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [9]:
# inspecting the imdb.title.crew dataset
titles_crew_df.head()

tconst                      directors              writers
0  tt0285252                      nm0899854            nm0899854
1  tt0438973                            NaN  nm0175726,nm1802864
2  tt0462036                      nm1940585            nm1940585
3  tt0835418                      nm0151540  nm0310087,nm0841532
4  tt0878654  nm0089502,nm2291498,nm2292011            nm0284943

In [10]:
# inspecting the imdb.title.principals dataset
titles_principals_df.head()

tconst  ordering     nconst  category       job        characters
0  tt0111414         1  nm0246005     actor       NaN       ["The Man"]
1  tt0111414         2  nm0398271  director       NaN               NaN
2  tt0111414         3  nm3739909  producer  producer               NaN
3  tt0323808        10  nm0059247    editor       NaN               NaN
4  tt0323808         1  nm3579312   actress       NaN  ["Beth Boothby"]

In [11]:
# inspecting the imdb.title.ratings.csv dataset
titles_ratings_df.head()

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [12]:
# inspecting the rt.movie_info.tsv dataset
movie_info_df.head()

synopsis rating  \
id                                                             
1   This gritty, fast-paced, and innovative police...      R   
3   New York City, not-too-distant-future: Eric Pa...      R   
5   Illeana Douglas delivers a superb performance ...      R   
6   Michael Douglas runs afoul of a treacherous su...      R   
7                                                 NaN     NR   

                                  genre          director  \
id                                                          
1   Action and Adventure|Classics|Drama  William Friedkin   
3     Drama|Science Fiction and Fantasy  David Cronenberg   
5     Drama|Musical and Performing Arts    Allison Anders   
6            Drama|Mystery and Suspense    Barry Levinson   
7                         Drama|Romance    Rodney Bennett   

                             writer  theater_date      dvd_date currency  \
id                                                                         
1                    Ernest Tidyman   Oct 9, 1971  Sep 25, 2001      NaN   
3      David Cronenberg|Don DeLillo  Aug 17, 2012   Jan 1, 2013        $   
5                    Allison Anders  Sep 13, 1996  Apr 18, 2000      NaN   
6   Paul Attanasio|Michael Crichton   Dec 9, 1994  Aug 27, 1997      NaN   
7                      Giles Cooper           NaN           NaN      NaN   

   box_office      runtime             studio  
id                                             
1         NaN  104 minutes                NaN  
3     600,000  108 minutes  Entertainment One  
5         NaN  116 minutes                NaN  
6         NaN  128 minutes                NaN  
7         NaN  200 minutes                NaN

In [13]:
# inspecting the rt.reviews.tsv dataset
reviews_df.head()

review rating   fresh  \
id                                                                     
3   A distinctly gallows take on contemporary fina...    3/5   fresh   
3   It's an allegory in search of a meaning that n...    NaN  rotten   
3   ... life lived in a bubble in financial dealin...    NaN   fresh   
3   Continuing along a line introduced in last yea...    NaN   fresh   
3              ... a perverse twist on neorealism...     NaN   fresh   

            critic  top_critic         publisher               date  
id                                                                   
3       PJ Nabarro           0   Patrick Nabarro  November 10, 2018  
3   Annalee Newitz           0           io9.com       May 23, 2018  
3     Sean Axmaker           0  Stream on Demand    January 4, 2018  
3    Daniel Kasman           0              MUBI  November 16, 2017  
3              NaN           0      Cinema Scope   October 12, 2017

In [14]:
# inspecting the tmdb.movies.csv dataset
tmdb_movies_df.head()

Unnamed: 0            genre_ids     id original_language  \
0           0      [12, 14, 10751]  12444                en   
1           1  [14, 12, 16, 10751]  10191                en   
2           2        [12, 28, 878]  10138                en   
3           3      [16, 35, 10751]    862                en   
4           4        [28, 878, 12]  27205                en   

                                 original_title  popularity release_date  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533   2010-11-19   
1                      How to Train Your Dragon      28.734   2010-03-26   
2                                    Iron Man 2      28.515   2010-05-07   
3                                     Toy Story      28.005   1995-11-22   
4                                     Inception      27.920   2010-07-16   

                                          title  vote_average  vote_count  
0  Harry Potter and the Deathly Hallows: Part 1           7.7       10788  
1                      How to Train Your Dragon           7.7        7610  
2                                    Iron Man 2           6.8       12368  
3                                     Toy Story           7.9       10174  
4                                     Inception           8.3       22186

In [15]:
# inspecting the tn.movie_budgets.csv dataset
tn_movie_budget_df.head()

release_date                                        movie  \
id                                                              
1   Dec 18, 2009                                       Avatar   
2   May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
3    Jun 7, 2019                                 Dark Phoenix   
4    May 1, 2015                      Avengers: Age of Ultron   
5   Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

   production_budget domestic_gross worldwide_gross  
id                                                   
1       $425,000,000   $760,507,625  $2,776,345,279  
2       $410,600,000   $241,063,875  $1,045,663,875  
3       $350,000,000    $42,762,350    $149,762,350  
4       $330,600,000   $459,005,868  $1,403,013,963  
5       $317,000,000   $620,181,382  $1,316,721,747

### Dataset Selection:

The following datasets were identified to contain features necessary to answer the formulated questions:

1. **Title Basics Dataset:**
   - Relevant features:
     - Primary title
     - Runtime in minutes
     - Genre

2. **Tn.movie_budgets Dataset:**
   - Contains production and box office performance data:
     - Movie name
     - Production budget
     - Domestic gross
     - Worldwide gross
     - Release Date

3. **Tmdb Movies Dataset:**
   - Relevant features:
     - Movie rating/Vote average
     - Numvotes
     - Popularity

4. **Movie Info Dataset:**
   - Relevant features:
     - Director
     - Writer
     - Rating

With these four datasets, we have sufficient information to answer the formulated questions.


### Further Inspection of selected datasets

In [17]:
# checking completeness and datatypes of the titles basics dataset
titles_basics_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   tconst           146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


In [16]:
# checking completeness and datatypes of the tn_movie_budget dataset
tn_movie_budget_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5782 entries, 1 to 82
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   release_date       5782 non-null   object
 1   movie              5782 non-null   object
 2   production_budget  5782 non-null   object
 3   domestic_gross     5782 non-null   object
 4   worldwide_gross    5782 non-null   object
dtypes: object(5)
memory usage: 271.0+ KB


In [18]:
# checking completeness and datatypes of the tmbd dataset
tmdb_movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26517 entries, 0 to 26516
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         26517 non-null  int64  
 1   genre_ids          26517 non-null  object 
 2   id                 26517 non-null  int64  
 3   original_language  26517 non-null  object 
 4   original_title     26517 non-null  object 
 5   popularity         26517 non-null  float64
 6   release_date       26517 non-null  object 
 7   title              26517 non-null  object 
 8   vote_average       26517 non-null  float64
 9   vote_count         26517 non-null  int64  
dtypes: float64(2), int64(3), object(5)
memory usage: 2.0+ MB


In [20]:
# checking completeness and datatypes of the movie info dataset
movie_info_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1560 entries, 1 to 2000
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   synopsis      1498 non-null   object
 1   rating        1557 non-null   object
 2   genre         1552 non-null   object
 3   director      1361 non-null   object
 4   writer        1111 non-null   object
 5   theater_date  1201 non-null   object
 6   dvd_date      1201 non-null   object
 7   currency      340 non-null    object
 8   box_office    340 non-null    object
 9   runtime       1530 non-null   object
 10  studio        494 non-null    object
dtypes: object(11)
memory usage: 146.2+ KB


In [21]:
titles_crew_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 3 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   tconst     146144 non-null  object
 1   directors  140417 non-null  object
 2   writers    110261 non-null  object
dtypes: object(3)
memory usage: 3.3+ MB


### Inspection Notes:

**1. Title Basics Dataset:**
- The dataset contains *146,144* entries. The `runtime_minutes` column is missing some values (has *114,405* non-null entries).

**Data Cleaning**
- To ensure accurate runtime analysis, we will retain entries where the `runtime_minutes` column is not null. For analyses that don't require runtime information, we can still include the entries.
- The `genre` column has compound objects, with some entries containing multiple genres separated by commas.
  - **Assumption:** For this analysis, we will consider the first genre listed in the `genre` column as the main genre, the second as the secondary genre, and the third as the tertiary genre.

**2. tn_movie_budget dataset:**
- The dataset contains *5,782* entries.
- The `budget`, `domestic_gross`, and `worldwide_gross` columns, which are expected to be of floating-point dtype, were found to be of dtype object.
- The `release_date` column has object dtype instead of datetime dtype.

**Data Cleaning**
- To facilitate further analysis, we will extract numerical values from the `budget`, `domestic_gross`, and `worldwide_gross` columns and convert them to float dtype.
- To enable date-based analysis, we will cast the `release_date` column from object type to datetime dtype.

**3. tmdb dataset:**
- The dataset contains *26,517* entries.
- All columns except the `release_date` column have the expected dtypes.

**Data Cleaning**
- To facilitate date-based analysis, we will cast the `release_date` column from object type to datetime dtype.

**4. movie_info dataset:**
- The dataset contains ~1,557 entries. Unfortunately, the dataset contains too few entries to be useful for our analysis. However, the `titles_crew` dataset has the director and writer information that was originally contained in the `movie_info` dataset.

**Data Cleaning**
- The `directors` and `writers` columns have compound entries, with some entries containing multiple strings separated by commas, indicating that a movie was directed or written by more than one individual.
  - **Assumption:** For this analysis, we will assume that the first director/writer listed in the column is the main director/writer, the second is considered the assistant director/writer, and the third is the supporting director/writer.

### Data Preparation: Cleaning & Wrangling

In summary, the inspection notes detail the key observations and data cleaning steps necessary to prepare the datasets for analysis. By addressing missing values, casting datatypes, and handling compound entries, we can ensure the data is accurate and suitable for further exploration and insights generation.


In [25]:
titles_basics_df.set_index('tconst').head(3)

primary_title              original_title  \
tconst                                                                   
tt0063540                        Sunghursh                   Sunghursh   
tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
tt0069049       The Other Side of the Wind  The Other Side of the Wind   

           start_year  runtime_minutes              genres  
tconst                                                      
tt0063540        2013            175.0  Action,Crime,Drama  
tt0066787        2019            114.0     Biography,Drama  
tt0069049        2018            122.0               Drama

In [24]:
titles_crew_df.head(3)

tconst  directors              writers
0  tt0285252  nm0899854            nm0899854
1  tt0438973        NaN  nm0175726,nm1802864
2  tt0462036  nm1940585            nm1940585

In [101]:
def split_compound_entries(dataframe, columns):
    '''
    Splits compound entries in multiple DataFrame columns into separate strings.

    Parameters:
        dataframe (pd.DataFrame): The DataFrame containing the columns to be processed.
        columns (list): A list of column names in the DataFrame to split compound entries.

    Returns:
        pd.DataFrame: A new DataFrame with additional columns containing the split strings.
    '''
    # Ensure all input columns exist in the DataFrame
    for column in columns:
        if column not in dataframe.columns:
            raise ValueError(f"Column '{column}' not found in the DataFrame.")
    
    # Initialize an empty dictionary to hold new split columns
    new_columns = {}
    
    # Process each input column
    for column in columns:
        # Get the index of the dataframe
        index = dataframe.index

        # Create new column names for the split strings
        col_1, col_2, col_3 = [(f"{column}_{i}") for i in range(1, 4)]
        
        # Initialize lists to hold the new split values
        new_col_1, new_col_2, new_col_3 = [], [], []

        # Iterate over each entry in the specified column
        for entry in dataframe[column]:
            try:
                # Split the entry using a comma as the separator
                items = entry.split(",") 

                # Keep only the first three items if there are more than three
                if len(items) > 3:
                    items = items[:3]
                
                # Assign the split items to the new columns accordingly
                if len(items) == 1:
                    new_col_1.append(items[0])
                    new_col_2.append(None)
                    new_col_3.append(None)
                elif len(items) == 2:
                    new_col_1.append(items[0])
                    new_col_2.append(items[1])
                    new_col_3.append(None)
                elif len(items) == 3:
                    new_col_1.append(items[0])
                    new_col_2.append(items[1])
                    new_col_3.append(items[2])
            except AttributeError:
                # Handle the case when the entry is not a string (e.g., NaN)
                new_col_1.append(None)
                new_col_2.append(None)
                new_col_3.append(None)
        
        # Store the new split values in the dictionary with column name as the key
        new_columns[col_1] = new_col_1
        new_columns[col_2] = new_col_2
        new_columns[col_3] = new_col_3
    
    # Create a new DataFrame from the new_columns dictionary
    df = pd.DataFrame(new_columns)
    
    # Drop the original columns from the DataFrame
    new_df = dataframe.drop(columns=columns)
    
    # Join the new DataFrame with the DataFrame containing the new split columns based on the index
    new_df = new_df.join(df, on=df.index)
    
    return new_df


In [102]:
# Split compound entries in the title_basics DataFrame for the 'genres' column
# and create new DataFrame nw_title_basics_df containing the split columns.
nw_title_basics_df = split_compound_entries(titles_basics_df.set_index('tconst'), ['genres'])

# Split compound entries in the title_crew DataFrame for the 'directors' and 'writers' columns
# and create new DataFrame nw_titles_crew_df containing the split columns.
nw_titles_crew_df = split_compound_entries(titles_crew_df.set_index('tconst'), ['directors', 'writers'])


#### Merging dataframes:

In [103]:
# We are joining these two DataFrames using the 'tconst' column as the common index
# to create a new combined DataFrame named 'movies_df'.
movies_df = nw_title_basics_df.join(nw_titles_crew_df, on=['tconst'])

# Displaying the first few rows of the combined DataFrame 'movies_df'.
movies_df.head()

primary_title              original_title  \
tconst                                                                   
tt0063540                        Sunghursh                   Sunghursh   
tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
tt0069049       The Other Side of the Wind  The Other Side of the Wind   
tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
tt0100275         The Wandering Soap Opera       La Telenovela Errante   

           start_year  runtime_minutes   genres_1 genres_2 genres_3  \
tconst                                                                
tt0063540        2013            175.0     Action    Crime    Drama   
tt0066787        2019            114.0  Biography    Drama     None   
tt0069049        2018            122.0      Drama     None     None   
tt0069204        2018              NaN     Comedy    Drama     None   
tt0100275        2017             80.0     Comedy    Drama  Fantasy   

          directors_1 directors_2 directors_3  writers_1  writers_2  writers_3  
tconst                                                                          
tt0063540   nm0712540        None        None  nm0023551  nm1194313  nm0347899  
tt0066787   nm0002411        None        None       None       None       None  
tt0069049   nm0000080        None        None  nm0000080  nm0462648       None  
tt0069204   nm0611531        None        None  nm0347899       None       None  
tt0100275   nm0765384   nm0749914        None  nm1360635  nm0749914       None

In [197]:
# inspecting the joined dataframe
movies_df.reset_index(inplace=True)

In [198]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 14 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   tconst           146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres_1         140736 non-null  object 
 6   genres_2         59378 non-null   object 
 7   genres_3         29436 non-null   object 
 8   directors_1      140417 non-null  object 
 9   directors_2      15727 non-null   object 
 10  directors_3      2501 non-null    object 
 11  writers_1        110261 non-null  object 
 12  writers_2        42229 non-null   object 
 13  writers_3        14383 non-null   object 
dtypes: float64(1), int64(1), object(12)
memory usage: 15.6+ MB


In [199]:
# inspecting the name basics dataset
name_basics_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 606648 entries, 0 to 606647
Data columns (total 6 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   nconst              606648 non-null  object 
 1   primary_name        606648 non-null  object 
 2   birth_year          82736 non-null   float64
 3   death_year          6783 non-null    float64
 4   primary_profession  555308 non-null  object 
 5   known_for_titles    576444 non-null  object 
dtypes: float64(2), object(4)
memory usage: 27.8+ MB


In [114]:
def add_name_columns(df, names_df, id_columns, name_prefix):
    """
    Add name columns to the DataFrame for directors and writers based on the provided 'names_df'.

    Parameters:
        df (pd.DataFrame): The main DataFrame to which name columns will be added.
        names_df (pd.DataFrame): The DataFrame containing IDs and names of directors and writers.
        id_columns (list): A list of columns in 'df' containing the IDs of directors and writers.
        name_prefix (str): A prefix to use for the name columns (e.g., 'director_name', 'writer_name').

    Returns:
        pd.DataFrame: A new DataFrame with the added name columns.
    """
    # Iterate through each ID column in 'id_columns'
    for id_column in id_columns:
        # Merge 'df' with 'names_df' to add the corresponding name column
        df = df.merge(names_df, left_on=id_column, right_on='nconst', how='left')
        
        # Rename the added 'primary_name' column to the specified name prefix and index (e.g., 'director_name_1')
        name_column = f"{name_prefix}_{id_column.split('_')[-1]}"
        df.rename(columns={'primary_name': name_column}, inplace=True)
        
        # Drop columns that are not needed
        df.drop(columns=['nconst', 'birth_year', 'death_year',
       'primary_profession', 'known_for_titles'], inplace=True)

    return df

In [200]:
# Define the columns in 'movies_df' that contain director IDs and writer IDs
director_id_columns = ['directors_1', 'directors_2', 'directors_3']
writer_id_columns = ['writers_1', 'writers_2', 'writers_3']

# Add name columns for directors using the 'add_name_columns' function
# 'new_movies_df' will be the DataFrame with added director name columns
new_movies_df = add_name_columns(movies_df, name_basics_df, director_id_columns, 'director_name')

# Add name columns for writers using the 'add_name_columns' function
# 'new_movies_df' will be the final DataFrame with both director and writer name columns
new_movies_df = add_name_columns(new_movies_df, name_basics_df, writer_id_columns, 'writer_name')


In [201]:
new_movies_df.head()

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes   genres_1 genres_2 genres_3 directors_1  \
0        2013            175.0     Action    Crime    Drama   nm0712540   
1        2019            114.0  Biography    Drama     None   nm0002411   
2        2018            122.0      Drama     None     None   nm0000080   
3        2018              NaN     Comedy    Drama     None   nm0611531   
4        2017             80.0     Comedy    Drama  Fantasy   nm0765384   

  directors_2 directors_3  writers_1  writers_2  writers_3  \
0        None        None  nm0023551  nm1194313  nm0347899   
1        None        None       None       None       None   
2        None        None  nm0000080  nm0462648       None   
3        None        None  nm0347899       None       None   
4   nm0749914        None  nm1360635  nm0749914       None   

        director_name_1 director_name_2 director_name_3 writer_name_1  \
0   Harnam Singh Rawail             NaN             NaN    Abrar Alvi   
1             Mani Kaul             NaN             NaN           NaN   
2          Orson Welles             NaN             NaN  Orson Welles   
3  Hrishikesh Mukherjee             NaN             NaN        Gulzar   
4     Valeria Sarmiento      Raoul Ruiz             NaN       Pía Rey   

    writer_name_2 writer_name_3  
0  Mahasweta Devi        Gulzar  
1             NaN           NaN  
2       Oja Kodar           NaN  
3             NaN           NaN  
4      Raoul Ruiz           NaN

In [207]:
# Convert the "release_date" column to datetime dtype in tn_movie_budget_df DataFrame
tn_movie_budget_df["release_date"] = pd.to_datetime(tn_movie_budget_df["release_date"])

# Extract the release year from the "release_date" column and create a new column "release_year"
tn_movie_budget_df["release_year"] = pd.to_datetime(tn_movie_budget_df["release_date"]).dt.year

# Extract the release month from the "release_date" column and create a new column "release_month"
tn_movie_budget_df["release_month"] = pd.to_datetime(tn_movie_budget_df["release_date"]).dt.month

# Extract the release day from the "release_date" column and create a new column "release_day"
tn_movie_budget_df["release_day"] = pd.to_datetime(tn_movie_budget_df["release_date"]).dt.day


In [208]:
# confirming addition of new columns
tn_movie_budget_df.head()

release_date                                        movie  \
id                                                             
1    2009-12-18                                       Avatar   
2    2011-05-20  Pirates of the Caribbean: On Stranger Tides   
3    2019-06-07                                 Dark Phoenix   
4    2015-05-01                      Avengers: Age of Ultron   
5    2017-12-15            Star Wars Ep. VIII: The Last Jedi   

   production_budget domestic_gross worldwide_gross  release_year  \
id                                                                  
1       $425,000,000   $760,507,625  $2,776,345,279          2009   
2       $410,600,000   $241,063,875  $1,045,663,875          2011   
3       $350,000,000    $42,762,350    $149,762,350          2019   
4       $330,600,000   $459,005,868  $1,403,013,963          2015   
5       $317,000,000   $620,181,382  $1,316,721,747          2017   

    release_month  release_day  
id                              
1              12           18  
2               5           20  
3               6            7  
4               5            1  
5              12           15

In [228]:
# Merge the 'tn_movie_budget_df' DataFrame with the 'new_movies_df' DataFrame based on specified columns
# The 'tn_movie_budget_df' is the left DataFrame, and the 'new_movies_df' is the right DataFrame.
# We perform an 'inner' join, meaning only the common rows based on the specified columns will be included in the result.
# The specified columns for the left DataFrame are 'movie' and 'release_year'.
# The specified columns for the right DataFrame are 'primary_title' and 'start_year'.

movies_performance_df = tn_movie_budget_df.merge(new_movies_df, left_on=['movie', 'release_year'], 
                              right_on=['primary_title', 'start_year'], how='inner')


In [235]:
# Merge the 'movies_performance_df' DataFrame with the 'titles_ratings_df' DataFrame based on the common column 'tconst'.
# The 'movies_performance_df' is the left DataFrame, and the 'titles_ratings_df' is the right DataFrame.
# We perform a 'left' join, meaning all rows from the left DataFrame will be included in the result, and only matching rows from the right DataFrame will be added.
# The 'tconst' column is used as the key for merging, as it is present in both DataFrames.

movies_performance_ratings_df = movies_performance_df.merge(titles_ratings_df, on='tconst', how='left')


#### Cleaning up duplicates

In [229]:
# Check for duplicates in the merged DataFrame based on the columns "movie" and "release_year"
duplicates_df = movies_performance_df[movies_performance_df.duplicated(subset=["movie", "release_year"])]

# Print the duplicate rows in the DataFrame
duplicates_df


release_date                movie production_budget domestic_gross  \
34     2010-03-05  Alice in Wonderland      $200,000,000   $334,191,110   
73     2017-11-22                 Coco      $175,000,000   $209,726,015   
148    2015-03-27                 Home      $130,000,000   $177,397,510   
149    2015-03-27                 Home      $130,000,000   $177,397,510   
150    2015-03-27                 Home      $130,000,000   $177,397,510   
...           ...                  ...               ...            ...   
1485   2015-06-23           Crossroads          $500,000             $0   
1513   2011-09-23              Weekend          $190,000       $484,592   
1514   2011-09-23              Weekend          $190,000       $484,592   
1529   2011-12-31             Absentia           $70,000             $0   
1543   2015-09-01               Exeter           $25,000             $0   

     worldwide_gross  release_year  release_month  release_day     tconst  \
34    $1,025,491,110          2010              3            5  tt2049386   
73      $798,008,101          2017             11           22  tt7002100   
148     $385,997,896          2015              3           27  tt5121236   
149     $385,997,896          2015              3           27  tt5775862   
150     $385,997,896          2015              3           27  tt5884960   
...              ...           ...            ...          ...        ...   
1485              $0          2015              6           23  tt5459744   
1513      $1,577,585          2011              9           23  tt1791679   
1514      $1,577,585          2011              9           23  tt1910670   
1529          $8,555          2011             12           31  tt7501544   
1543        $489,792          2015              9            1  tt4931374   

            primary_title  ... directors_3  writers_1  writers_2 writers_3  \
34    Alice in Wonderland  ...        None       None       None      None   
73                   Coco  ...        None  nm9039216       None      None   
148                  Home  ...        None       None       None      None   
149                  Home  ...        None       None       None      None   
150                  Home  ...        None  nm8289285  nm7057993      None   
...                   ...  ...         ...        ...        ...       ...   
1485           Crossroads  ...        None       None       None      None   
1513              Weekend  ...        None  nm4214252       None      None   
1514              Weekend  ...        None  nm3248760       None      None   
1529             Absentia  ...        None  nm9339179       None      None   
1543               Exeter  ...        None       None       None      None   

       director_name_1 director_name_2 director_name_3       writer_name_1  \
34    James Fotopoulos             NaN             NaN                 NaN   
73        Neil Boultby             NaN             NaN        Neil Boultby   
148           Ryan Lee             NaN             NaN                 NaN   
149         Ben Ducker             NaN             NaN                 NaN   
150    Asaf Lavi Harel             NaN             NaN     Oren Cahanovich   
...                ...             ...             ...                 ...   
1485    Junichi Suzuki             NaN             NaN                 NaN   
1513     Cezary Pazura             NaN             NaN  Leslaw Kazmierczak   
1514    Randal Kamradt             NaN             NaN      Randal Kamradt   
1529      Steven Young             NaN             NaN        Steven Young   
1543               NaN             NaN             NaN                 NaN   

        writer_name_2 writer_name_3  
34                NaN           NaN  
73                NaN           NaN  
148               NaN           NaN  
149               NaN           NaN  
150   Asaf Lavi Harel           NaN  
...               ...           ...  
1485              NaN           NaN  
1

In [230]:
# Drop the duplicate rows based on the columns "movie" and "release_year"
movies_performance_df.drop_duplicates(subset=["movie", "release_year"], inplace=True)



<class 'pandas.core.frame.DataFrame'>
Index: 1547 entries, tt1298650 to tt3973612
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   release_date       1547 non-null   datetime64[ns]
 1   movie              1547 non-null   object        
 2   production_budget  1547 non-null   object        
 3   domestic_gross     1547 non-null   object        
 4   worldwide_gross    1547 non-null   object        
 5   release_year       1547 non-null   int64         
 6   release_month      1547 non-null   int64         
 7   release_day        1547 non-null   int64         
 8   primary_title      1547 non-null   object        
 9   original_title     1547 non-null   object        
 10  start_year         1547 non-null   int64         
 11  runtime_minutes    1521 non-null   float64       
 12  genres_1           1541 non-null   object        
 13  genres_2           1343 non-null   object        
 14  

In [231]:
# confirming duplicates have been deleted
duplicates_df = movies_performance_df[movies_performance_df.duplicated(subset=["movie", "release_year"])]

# Print the duplicate rows in the DataFrame
duplicates_df

Empty DataFrame
Columns: [release_date, movie, production_budget, domestic_gross, worldwide_gross, release_year, release_month, release_day, tconst, primary_title, original_title, start_year, runtime_minutes, genres_1, genres_2, genres_3, directors_1, directors_2, directors_3, writers_1, writers_2, writers_3, director_name_1, director_name_2, director_name_3, writer_name_1, writer_name_2, writer_name_3]
Index: []

[0 rows x 28 columns]

In [236]:
movies_performance_ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1484 entries, 0 to 1483
Data columns (total 30 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   release_date       1484 non-null   datetime64[ns]
 1   movie              1484 non-null   object        
 2   production_budget  1484 non-null   object        
 3   domestic_gross     1484 non-null   object        
 4   worldwide_gross    1484 non-null   object        
 5   release_year       1484 non-null   int64         
 6   release_month      1484 non-null   int64         
 7   release_day        1484 non-null   int64         
 8   tconst             1484 non-null   object        
 9   primary_title      1484 non-null   object        
 10  original_title     1484 non-null   object        
 11  start_year         1484 non-null   int64         
 12  runtime_minutes    1467 non-null   float64       
 13  genres_1           1484 non-null   object        
 14  genres_2

In [237]:
movies_performance_ratings_df.head()

release_date                                        movie production_budget  \
0   2011-05-20  Pirates of the Caribbean: On Stranger Tides      $410,600,000   
1   2019-06-07                                 Dark Phoenix      $350,000,000   
2   2015-05-01                      Avengers: Age of Ultron      $330,600,000   
3   2018-04-27                       Avengers: Infinity War      $300,000,000   
4   2017-11-17                               Justice League      $300,000,000   

  domestic_gross worldwide_gross  release_year  release_month  release_day  \
0   $241,063,875  $1,045,663,875          2011              5           20   
1    $42,762,350    $149,762,350          2019              6            7   
2   $459,005,868  $1,403,013,963          2015              5            1   
3   $678,815,482  $2,048,134,200          2018              4           27   
4   $229,024,295    $655,945,209          2017             11           17   

      tconst                                primary_title  ...  writers_2  \
0  tt1298650  Pirates of the Caribbean: On Stranger Tides  ...  nm0744429   
1  tt6565702                                 Dark Phoenix  ...  nm1079208   
2  tt2395427                      Avengers: Age of Ultron  ...  nm0498278   
3  tt4154756                       Avengers: Infinity War  ...  nm1321656   
4  tt0974015                               Justice League  ...  nm0795975   

   writers_3  director_name_1 director_name_2 director_name_3  \
0  nm0064181     Rob Marshall             NaN             NaN   
1  nm1079211    Simon Kinberg             NaN             NaN   
2  nm0456158      Joss Whedon             NaN             NaN   
3  nm0498278    Anthony Russo       Joe Russo             NaN   
4  nm0006516      Zack Snyder             NaN             NaN   

        writer_name_1    writer_name_2   writer_name_3 averagerating  numvotes  
0         Ted Elliott     Terry Rossio  Stuart Beattie           6.6  447624.0  
1          John Byrne  Chris Claremont    Dave Cockrum           6.0   24451.0  
2         Joss Whedon         Stan Lee      Jack Kirby           7.3  665594.0  
3  Christopher Markus  Stephen McFeely        Stan Lee           8.5  670926.0  
4        Jerry Siegel      Joe Shuster    Chris Terrio           6.5  329135.0  

[5 rows x 30 columns]

### Feature Engineering:
The ROI (Return on Investment) is a crucial feature in the dataset that measures the financial performance of a movie. It indicates the profitability of a movie by calculating the net income as a ratio of the production budget. Below are some important points related to ROI in the context of the given dataset:

Calculation of ROI:

ROI is calculated as the net movie income divided by the production budget.
The net movie income can be derived by subtracting the production budget from the worldwide gross (i.e., total earnings from domestic and international box office).
The production budget, domestic gross, and worldwide gross columns should be converted to numeric (float) data types before calculating ROI.

`Worldwide_ROI = (worldwide_gross - production_budget) / production_budget`

`Domestic_ROI = (domestic_gross - production_budget) / production_budget`
- **Data Cleaning and Conversion:**

The columns 'production_budget', 'domestic_gross', and 'worldwide_gross' are currently of object data type. To calculate ROI, these columns should be converted to numeric (float) data types.

In [240]:
def convert_object_to_float(dataframe, columns):
    """
    Convert specified columns from object dtype to float dtype.

    Parameters: 
        dataframe (pd.DataFrame): The DataFrame whose columns will be changed.
        columns (list): List of column names to cast from object to float.

    Returns:
        pd.DataFrame: The DataFrame with specified columns converted to float dtype.
    """
    # Iterate through the list of columns to convert
    for column in columns:
        # Remove '$' and ',' from the values in the column, then convert to float
        dataframe[column] = dataframe[column].str.replace("$", "").str.replace(",", "").astype(float)
    
    return dataframe


In [241]:
# convert production_budget, worldwide_gross, and domestic_gross columns from object dtype to float dtype
columns_to_convert = ["production_budget", "worldwide_gross", "domestic_gross"]

movies_performance_ratings_df = convert_object_to_float(movies_performance_ratings_df, columns_to_convert)

In [242]:
# confirm conversion production_budget, worldwide_gross, and domestic_gross columns from object dtype to float dtype 
movies_performance_ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1484 entries, 0 to 1483
Data columns (total 30 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   release_date       1484 non-null   datetime64[ns]
 1   movie              1484 non-null   object        
 2   production_budget  1484 non-null   float64       
 3   domestic_gross     1484 non-null   float64       
 4   worldwide_gross    1484 non-null   float64       
 5   release_year       1484 non-null   int64         
 6   release_month      1484 non-null   int64         
 7   release_day        1484 non-null   int64         
 8   tconst             1484 non-null   object        
 9   primary_title      1484 non-null   object        
 10  original_title     1484 non-null   object        
 11  start_year         1484 non-null   int64         
 12  runtime_minutes    1467 non-null   float64       
 13  genres_1           1484 non-null   object        
 14  genres_2

#### ROI Calculation: Domestic ROI & Worldwide ROI

In [243]:
# Calculate Domestic ROI (Return on Investment)
movies_performance_ratings_df["domestic_roi"] = (movies_performance_ratings_df["domestic_gross"] - movies_performance_ratings_df['production_budget']) / movies_performance_ratings_df["production_budget"]

# Calculate Worldwide ROI (Return on Investment)
movies_performance_ratings_df["worldwide_roi"] = (movies_performance_ratings_df["worldwide_gross"] - movies_performance_ratings_df['production_budget']) / movies_performance_ratings_df["production_budget"]


In [244]:
movies_performance_ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1484 entries, 0 to 1483
Data columns (total 32 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   release_date       1484 non-null   datetime64[ns]
 1   movie              1484 non-null   object        
 2   production_budget  1484 non-null   float64       
 3   domestic_gross     1484 non-null   float64       
 4   worldwide_gross    1484 non-null   float64       
 5   release_year       1484 non-null   int64         
 6   release_month      1484 non-null   int64         
 7   release_day        1484 non-null   int64         
 8   tconst             1484 non-null   object        
 9   primary_title      1484 non-null   object        
 10  original_title     1484 non-null   object        
 11  start_year         1484 non-null   int64         
 12  runtime_minutes    1467 non-null   float64       
 13  genres_1           1484 non-null   object        
 14  genres_2

In [245]:
!pwd

/c/Users/user/Documents/Flatiron/dsc-movie-studio-project


In [246]:
# specify the path and filename to save the CSV file
output_path = r"C:\Users\user\Documents\Flatiron\dsc-movie-studio-project\data\cleaned_movies.csv"

# Export the DataFrame CSV file
movies_performance_ratings_df.to_csv(output_path, index=False)
